distribution/pull_v2.go

```diff
@@ -182,13 +188,25 @@ func (ld *v2LayerDescriptor) Download(ctx context.Context, progressOutput progre
 	tmpFile := ld.tmpFile
 	blobs := ld.repo.Blobs(ctx)
 
-	layerDownload, err := blobs.Open(ctx, ld.digest)
-	if err != nil {
-		logrus.Errorf("Error initiating layer download: %v", err)
-		if err == distribution.ErrBlobUnknown {
-			return nil, 0, xfer.DoNotRetry{Err: err}
+	var layerDownload distribution.ReadSeekCloser
+	if ld.foreignSrc == nil {
+		layerDownload, err = blobs.Open(ctx, ld.digest)
+		if err != nil {
+			logrus.Errorf("Error initiating layer download: %v", err)
+			if err == distribution.ErrBlobUnknown {
+				return nil, 0, xfer.DoNotRetry{Err: err}
+			}
+			return nil, 0, retryOnError(err)
+		}
+	} else {
+		// Find the first URL that results in a 200 result code.
+		for _, url := range ld.foreignSrc.URLs {
+			layerDownload = transport.NewHTTPReadSeeker(http.DefaultClient, url, nil)
+			_, err = layerDownload.Seek(0, os.SEEK_SET)
+			if err == nil {
+				break
+			}
```

Comment:
Call `layerDownload.Close()` here to avoid leaking connections associated with unsuccessful attempts.